### ISSUE TITLE
tracing batch normalization: apparently unnecessary gpu-cpu-gpu transfer for gather op.

### ISSUE THREAD

**Issue by liuyipei** [tensorflow/tensorflow#6027]
I am running:
CUDA 8.0. CUDNN 5.1. tensorflow 0.11. Ubuntu 12.04.

I was experiencing low GPU utilization. As per #1824 I did a trace and found many instances of the following interaction, as shown by the screenshot in chrome:

The op I have highlighted (with arrows going in and out in the screenshot) is a `gather` op. I provide a link to the context in which it is called.
https://github.com/tensorflow/tensorflow/blob/eb56a8af24695bf8258addf28b0c53fbabff72e6/tensorflow/python/ops/nn_impl.py#L472

The call stack (at graph construction) time is:
`tf.contrib.layers.batch_norm`, `python.ops.nn_impl.moments`, `python.ops.nn_impl.sufficient_statics`

If I understand correctly, I am seeing that the GPU gets blocked as a result of shuffling a trivial `gather` op unnecessarily off to the CPU only to wait forever for it to come back. 

![screenshot from 2016-12-01 11 45 17](https://cloud.githubusercontent.com/assets/4404828/20810886/67cbd92e-b7c0-11e6-991d-75ab54a1c699.png)

**Comment by poxvoculi** [comment]
The gather Op is being placed on CPU because all of its arguments are int32.  (This is fallout from an early mis-design that hasn't yet been completely remediated.)  In this case, the gather is rather trivial and probably shouldn't be used.  @vincentvanhoucke

**Comment by liuyipei** [comment]
Sorry to press the issue. @poxvoculi  In this case, what do you think is the best way to address it?

**Comment by vincentvanhoucke** [comment]
I'll take a look. If anyone has a solution handy, send me a PR.

**Comment by poxvoculi** [comment]
Oops. Closed by mistake.  The python in sufficient_statistics should be rewritten to avoid using array_ops.gather in this case.

**Comment by vincentvanhoucke** [comment]
@liuyipei if you confirm that #6040 fixes the problem, then my fix at #6031 is slightly cleaner and should be equivalent.

**Comment by liuyipei** [comment]
@vincentvanhoucke Thank you very much! I have closed my PR.
